plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/SearchQueryRecord.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.rules.model;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.search.SearchType;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+import org.opensearch.core.xcontent.ToXContentObject;
+import org.opensearch.core.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Simple abstract class that represent record stored in the Query Insight Framework
+ *
+ * @param <T> The value type associated with the record
+ * @opensearch.internal
+ */
+public abstract class SearchQueryRecord<T extends Number & Comparable<T>>
+    implements
+        Comparable<SearchQueryRecord<T>>,
+        Writeable,
+        ToXContentObject {
+
+    private static final Logger log = LogManager.getLogger(SearchQueryRecord.class);
+    protected static final String TIMESTAMP = "timestamp";
+    protected static final String SEARCH_TYPE = "searchType";
+    protected static final String SOURCE = "source";
+    protected static final String TOTAL_SHARDS = "totalShards";
+    protected static final String INDICES = "indices";
+    protected static final String PROPERTY_MAP = "propertyMap";
+    protected static final String VALUE = "value";
```

Comment:
Should be private as properly designed need not worry about these serialization strings